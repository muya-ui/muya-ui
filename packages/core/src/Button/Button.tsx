import React from 'react';
import styled, { css, CSSObject, ThemedStyledProps } from 'styled-components';

import { colorUtils, ITheme } from '@muya-ui/theme-light';

import { withThemeForStyled } from '../utils/withTheme';
import ButtonCore from './ButtonCore';
import { buttonSizeCSS, highlightNeutralColorButtonType, siblingCss } from './mixins';
import { IButtonNode, IButtonProps, IButtonType } from './types';

const ButtonPure = React.forwardRef<IButtonNode, IButtonProps>((props, ref) => {
  const { type, shape, block, groupType, offBorder, plain, selected, ...otherProps } = props;

  return <ButtonCore ref={ref} {...otherProps} />;
});

const buttonInGroupCss = (groupType: IButtonProps['groupType']) => {
  return css`
    ${groupType === 'head' &&
      css`
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      `}

    ${groupType === 'tail' &&
      css`
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        margin-left: -1px;
      `}
  `;
};

interface IButtonCSSConfig {
  style: CSSObject;
  hoverStyle?: CSSObject;
  clickStyle?: CSSObject;
}
export function mergeCSSConfig(a: IButtonCSSConfig, b: IButtonCSSConfig) {
  return {
    style: {
      ...a.style,
      ...b.style,
    },
    hoverStyle: {
      ...a.hoverStyle,
      ...b.hoverStyle,
    },
    clickStyle: {
      ...a.clickStyle,
      ...b.clickStyle,
    },
  };
}

const buttonBorderCSS = (
  theme: ITheme,
  plain?: boolean,
  groupType?: IButtonProps['groupType'],
  offBorder?: IButtonProps['offBorder'],
  busyOrLoading?: boolean,
) => {
  const { plainIsLight } = theme.components.Button;

  const groupCSSObj: CSSObject = {};
  // 面性的按钮在 ButtonGroup 中需要设置 border
  if ((!plain || plainIsLight) && groupType) {
    if (groupType !== 'tail') {
      groupCSSObj.borderRightWidth = '1px';
      groupCSSObj.borderRightStyle = 'solid';
    }

    if (groupType !== 'head') {
      groupCSSObj.borderLeftWidth = '1px';
      groupCSSObj.borderLeftStyle = 'solid';
    }
  }

  // 在 busy 或者 loading 时整体会有透明度，导致中间的border颜色变深
  // 需要处理一下
  if (busyOrLoading) {
    if (offBorder === 'both' || offBorder === 'right') {
      groupCSSObj.borderRightColor = 'transparent';
    }

    if (offBorder === 'both' || offBorder === 'left') {
      groupCSSObj.borderLeftColor = 'transparent';
    }
  }

  const noBorderGroupCSS = css(groupCSSObj);

  let borderCss;
  if (plain && !plainIsLight) {
    borderCss = css`
      border-style: solid;
      border-width: 1px;
    `;
  } else {
    borderCss = css`
      border-width: 0;
    `;
  }

  return css`
    ${borderCss}
    ${noBorderGroupCSS}
  `;
};

/* eslint-disable complexity */
function getOneTypeOfButton(
  theme: ITheme,
  type: IButtonType,
  interactive: boolean,
  plain?: boolean,
): IButtonCSSConfig {
  const {
    typeColor,
    defaultColor,
    typeBgImage,
    statusOpacity,
    textColor,
    lightTypeColor,
    plainIsLight,
    plainColorHighlight,
    borderColor: typeBorderColor,
    typeBgColor,
  } = theme.components.Button;
  const currentBgImage = typeBgImage[type];
  const currentColor = typeColor[type];
  const currentLightColor = lightTypeColor[type];
  const currentBgColor = typeBgColor[type];
  let currentTypeUseBgImage = !!currentBgImage;

  let borderColor;

  // 当是面性的按钮时，在 ButtonGroup 中是有 border 的
  // 需要设置 borderColor
  if (plain && plainIsLight) {
    borderColor = currentLightColor.hover;
  } else if (plain) {
    borderColor = typeBorderColor[type];
  } else {
    borderColor = currentColor.hover;
  }

  let color: string;
  if (plain) {
    color = currentColor.normal;
  } else if (textColor[type]) {
    color = textColor[type]!;
  } else {
    color = defaultColor;
  }

  let bgColor;
  if (plain && plainIsLight) {
    bgColor = currentLightColor.normal;
  } else if (plain) {
    bgColor = 'transparent';
  } else if (currentBgColor) {
    bgColor = currentBgColor.normal;
  } else {
    bgColor = currentColor.normal;
  }

  let hoverColor;
  let hoverBackgroundImage;
  let hoverBackgroundColor;
  let hoverBorderColor;

  let clickColor;
  let clickBackgroundImage;
  let clickBackgroundColor;
  let clickBorderColor;

  if (interactive) {
    if (plain && !plainIsLight) {
      hoverBackgroundColor = colorUtils.transparentize(statusOpacity.hover, color);
      clickBackgroundColor = colorUtils.transparentize(statusOpacity.click, color);
    } else if (plain) {
      hoverBackgroundColor = currentLightColor.hover;
      clickBackgroundColor = currentLightColor.click;
    } else if (currentTypeUseBgImage) {
      hoverBackgroundImage = currentBgImage!.hover;
      clickBackgroundImage = currentBgImage!.click;
    } else if (currentBgColor) {
      hoverBackgroundColor = currentBgColor.hover;
      clickBackgroundColor = currentBgColor.click;
    } else {
      hoverBackgroundColor = currentColor.hover;
      clickBackgroundColor = currentColor.click;
    }

    if (plain && plainColorHighlight && highlightNeutralColorButtonType(type)) {
      hoverColor = typeColor.primary.hover;
      clickColor = typeColor.primary.click;
      hoverBorderColor = typeColor.primary.hover;
      clickBorderColor = typeColor.primary.click;
    } else {
      hoverColor = color;
      clickColor = color;
    }
  }

  let backgroundImage;
  let backgroundColor;
  // 不是线框的按钮，同时使用背景
  if (!plain && currentTypeUseBgImage) {
    backgroundImage = currentBgImage!.normal;
  } else {
    backgroundColor = bgColor;
  }

  let busyOpacity;
  if (!interactive) {
    busyOpacity = theme.opacity.pattern.disabled;
  }

  return {
    style: {
      color,
      backgroundImage,
      backgroundColor,

      opacity: busyOpacity,
      cursor: interactive ? 'pointer' : 'not-allowed',
      borderColor,
    },
    hoverStyle: {
      color: hoverColor,
      backgroundImage: hoverBackgroundImage,
      backgroundColor: hoverBackgroundColor,

      borderColor: hoverBorderColor || borderColor,
    },
    clickStyle: {
      color: clickColor,
      backgroundImage: clickBackgroundImage,
      backgroundColor: clickBackgroundColor,

      borderColor: clickBorderColor || borderColor,
    },
  };
}

/* eslint-enable complexity */

function getDisabledButtonConfig(theme: ITheme, plain?: boolean): IButtonCSSConfig {
  const { defaultColor, plainIsLight } = theme.components.Button;
  const disabledColor = theme.colors.pattern.text.disabled;

  let color = defaultColor;
  let borderColor;
  let backgroundColor;

  // 如果是线框的
  if (plain && !plainIsLight) {
    color = disabledColor;
    borderColor = disabledColor;
    backgroundColor = 'transparent';
  } else {
    borderColor = defaultColor;
    backgroundColor = disabledColor;
  }

  return {
    style: {
      cursor: 'not-allowed',
      color,
      backgroundColor,
      borderColor,
    },
  };
}

function getButtonConfigInGroup(
  theme: ITheme,
  type: IButtonType,
  selected?: boolean,
): IButtonCSSConfig {
  const { hoverZIndex, selectedZIndex, zIndexOrder } = theme.components.Button;
  const currentZIndex = zIndexOrder[type];
  let activeZIndex;
  if (selected) {
    activeZIndex = selectedZIndex;
  } else {
    activeZIndex = hoverZIndex;
  }

  let normalZIndex;
  if (currentZIndex && selected) {
    normalZIndex = selectedZIndex;
  } else if (currentZIndex) {
    normalZIndex = currentZIndex;
  }
  return {
    style: {
      zIndex: normalZIndex,
    },
    clickStyle: {
      zIndex: activeZIndex,
    },
    hoverStyle: {
      zIndex: activeZIndex,
    },
  };
}

// 状态的样式
function buttonCSS(props: ThemedStyledProps<IButtonProps, ITheme>) {
  const {
    size = 'm',
    type = 'normal',
    plain,
    theme,
    busy,
    disabled,
    loading,
    groupType,
    offBorder,
    selected,
    disableSiblingMargin = false,
  } = props;
  const interactive = !(busy || disabled || loading);
  const baseConfig = disabled
    ? getDisabledButtonConfig(theme, plain)
    : getOneTypeOfButton(theme, type, interactive, plain);
  let config;
  if (groupType) {
    const groupConfig = getButtonConfigInGroup(theme, type, selected);
    config = mergeCSSConfig(baseConfig, groupConfig);
  } else {
    config = baseConfig;
  }

  const normalCss = css(config.style);
  let interactiveCss;
  if (interactive && config.hoverStyle && config.clickStyle) {
    interactiveCss = css`
      &:hover {
        ${css(config.hoverStyle)}
      }
      &:active {
        ${css(config.clickStyle)}
      }
    `;
  }
  const busyOrLoading = busy || loading;

  return css`
    ${normalCss}
    ${interactiveCss}
    ${buttonBorderCSS(theme, plain, groupType, offBorder, busyOrLoading)}
    ${siblingCss(size, theme, !!groupType, disableSiblingMargin)}
  `;
}

const Button = styled(ButtonPure)<IButtonProps>`
  ${props => {
    const { type, selected, groupType, plain, textFine, fontWeight } = props;
    let innerPlain = plain;
    let innerType = type;
    if (!innerType && selected === true) {
      innerType = 'primary';
    } else if (!innerType && selected === false) {
      innerType = 'secondary';
    } else if (!innerType) {
      innerType = 'normal';
    }
    if (innerType === 'normal' && plain === undefined) {
      innerPlain = true;
    }
    let innerFontWeight = fontWeight;
    if (!innerFontWeight && textFine) {
      innerFontWeight = 'lighter';
    }
    const newProps = {
      ...props,
      type: innerType,
      plain: innerPlain,
      fontWeight: innerFontWeight,
    };
    return css`
      user-select: none;
      ${buttonSizeCSS(newProps)}
      ${buttonInGroupCss(groupType)}
      ${buttonCSS(newProps)}
    `;
  }}
`;
(Button as any).__MUYA_BUTTON = true;
export default withThemeForStyled(Button);
