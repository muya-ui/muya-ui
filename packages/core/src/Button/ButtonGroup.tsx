import React from 'react';
import styled, { css } from 'styled-components';

import useTheme from '../utils/useTheme';
import { withThemeForStyled } from '../utils/withTheme';
import { siblingCss } from './mixins';
import { IButtonGroupProps, IButtonProps, IButtonType } from './types';

const BaseNode = styled.div``;
export interface IChildConfig extends Pick<IButtonProps, 'groupType' | 'offBorder'> {
  zIndex?: number;
}
export type IChildConfigMap = Record<string | number, IChildConfig>;

/* eslint-disable complexity */
/**
 * 在 Group 中需要根据 z-index 来决定谁的 border 在上面
 * @param childIndex
 * @param configMap
 * @param disabled
 */
export function getOffBorder(childIndex: number, configMap: IChildConfigMap, disabled?: boolean) {
  const { groupType, zIndex } = configMap[childIndex];
  let offBorder: IButtonProps['offBorder'];
  const prevItem = configMap[childIndex - 1];
  const nextItem = configMap[childIndex + 1];
  if (disabled && (groupType === 'head' || groupType === 'group')) {
    offBorder = 'right';
  } else if (disabled && groupType === 'tail') {
    offBorder = 'no';
  }
  // 是第一个的时候，有两种情况
  else if (groupType === 'head' && nextItem && nextItem.zIndex! < zIndex!) {
    offBorder = 'no';
  } else if (groupType === 'head') {
    offBorder = 'right';
  }
  // 是最后的一个的时候，也有两种情况
  else if (groupType === 'tail' && prevItem && prevItem.zIndex! > zIndex!) {
    offBorder = 'left';
  } else if (groupType === 'tail') {
    offBorder = 'no';
  }
  // 在中间的时候，有四种情况
  else if (groupType === 'group' && nextItem.zIndex! > zIndex! && prevItem.zIndex! > zIndex!) {
    offBorder = 'both';
  } else if (groupType === 'group' && nextItem.zIndex! < zIndex! && prevItem.zIndex! < zIndex!) {
    offBorder = 'no';
  } else if (groupType === 'group' && nextItem.zIndex! < zIndex! && prevItem.zIndex! > zIndex!) {
    offBorder = 'left';
  } else {
    offBorder = 'right';
  }
  return offBorder;
}
/* eslint-enable complexity */

const ButtonGroupWithoutStyle = React.memo(
  React.forwardRef<HTMLDivElement, IButtonGroupProps>((props, ref) => {
    const {
      type,
      size = 'm',
      htmlType,
      disabled,
      busy,
      shape,
      block,
      children,
      plain,
      ...otherProps
    } = props;
    const theme = useTheme();
    const childrenLen = React.Children.count(children);
    const notOnlyOne = childrenLen > 1;
    let resultChildren;
    const { zIndexOrder } = theme.components.Button;
    if (children) {
      const configMap: IChildConfigMap = {};
      React.Children.forEach(children, (child, childIndex) => {
        if (!React.isValidElement(child)) {
          return;
        }
        const currentType = (child.props.type as IButtonType) || type;
        let groupType: IButtonProps['groupType'];
        if (notOnlyOne && !childIndex) {
          groupType = 'head';
        } else if (notOnlyOne && childIndex === childrenLen - 1) {
          groupType = 'tail';
        } else if (notOnlyOne) {
          groupType = 'group';
        }

        configMap[childIndex] = {
          groupType,
          zIndex: zIndexOrder[currentType] || 0,
        };
      });
      resultChildren = React.Children.map(children, (child, childIndex) => {
        if (!React.isValidElement(child)) {
          return null;
        }
        const {
          disabled: childDisabled,
          busy: childBusy,
          shape: childShape,
          size: childSize,
          block: childBlock,
          type: childType,
          plain: childPlain,
          offBorder: childOffBorder,
          ...otherChildProps
        } = child.props;
        // 布尔值需要注意一下
        const currentDisabled = childDisabled !== undefined ? childDisabled : disabled;
        const offBorder: IButtonProps['offBorder'] =
          childOffBorder || getOffBorder(childIndex, configMap, currentDisabled);
        const currentType = childType || type;
        const { groupType } = configMap[childIndex];

        return React.cloneElement(child, {
          ...otherChildProps,
          disabled: currentDisabled,
          busy: childBusy !== undefined ? childBusy : busy,
          shape: childShape || shape,
          plain: childPlain !== undefined ? childPlain : plain,
          type: currentType,
          offBorder,
          groupType,
          size,
          block,
        });
      });
    }
    return (
      <BaseNode ref={ref} {...otherProps}>
        {resultChildren}
      </BaseNode>
    );
  }),
);

const ButtonGroup = styled(ButtonGroupWithoutStyle)`
  ${props => {
    const { size = 'm', block, theme, disableSiblingMargin = false } = props;
    return css`
      /* 相对定位 */
      position: relative;

      /* flex + inline */
      display: inline-flex;
      ${block && 'width: 100%;'}
      ${siblingCss(size, theme, false, disableSiblingMargin)}
    `;
  }}
`;

export default withThemeForStyled(ButtonGroup);
