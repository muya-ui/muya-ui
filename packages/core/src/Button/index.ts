export * from './types';
export { default as ButtonGroup } from './ButtonGroup';
export { default as Button } from './Button';
export { default as InlineButton } from './InlineButton';
export { default as MaskButton } from './MaskButton';
export { default } from './Button';
