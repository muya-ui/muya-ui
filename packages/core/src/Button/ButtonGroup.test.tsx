import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';
import ButtonGroup, { getOffBorder, IChildConfigMap } from './ButtonGroup';

test('测试 ButtonGroup 正常情况', () => {
  const tree = renderer
    .create(
      <ButtonGroup>
        <Button>AAA</Button>
        <Button>AAA</Button>
      </ButtonGroup>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 ButtonGroup 子元素属性传递', () => {
  const tree = renderer
    .create(
      <ButtonGroup size="xl" block>
        sss
        <Button>AAA</Button>
        <Button>AAA</Button>
        <Button>AAA</Button>
      </ButtonGroup>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 ButtonGroup 空情况', () => {
  const tree = renderer.create(<ButtonGroup size="xl" block></ButtonGroup>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 ButtonGroup 单节点', () => {
  const tree = renderer
    .create(
      <ButtonGroup size="xl" block>
        <Button>AAA</Button>
      </ButtonGroup>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 ButtonGroup selected 快捷方式', () => {
  const tree = renderer
    .create(
      <ButtonGroup>
        <Button selected={false}>AAA</Button>
        <Button selected>AAA</Button>
        <Button selected={false}>AAA</Button>
      </ButtonGroup>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 ButtonGroup getOffBorder', () => {
  const c1: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 3,
    },
    1: {
      groupType: 'head',
      zIndex: 2,
    },
  };
  expect(getOffBorder(0, c1)).toBe('no');
  const c2: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 3,
    },
    1: {
      groupType: 'head',
      zIndex: 4,
    },
  };
  expect(getOffBorder(0, c2)).toBe('right');
  const c3: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 3,
    },
    1: {
      groupType: 'tail',
      zIndex: 2,
    },
  };
  expect(getOffBorder(1, c3)).toBe('left');
  const c4: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 3,
    },
    1: {
      groupType: 'tail',
      zIndex: 4,
    },
  };
  expect(getOffBorder(1, c4)).toBe('no');
  const c5: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 3,
    },
    1: {
      groupType: 'group',
      zIndex: 2,
    },
    2: {
      groupType: 'tail',
      zIndex: 4,
    },
  };
  expect(getOffBorder(1, c5)).toBe('both');
  const c6: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 3,
    },
    1: {
      groupType: 'group',
      zIndex: 10,
    },
    2: {
      groupType: 'tail',
      zIndex: 4,
    },
  };
  expect(getOffBorder(1, c6)).toBe('no');
  const c7: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 3,
    },
    1: {
      groupType: 'group',
      zIndex: 4,
    },
    2: {
      groupType: 'tail',
      zIndex: 5,
    },
  };
  expect(getOffBorder(1, c7)).toBe('right');
  const c8: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 5,
    },
    1: {
      groupType: 'group',
      zIndex: 4,
    },
    2: {
      groupType: 'tail',
      zIndex: 3,
    },
  };
  expect(getOffBorder(1, c8)).toBe('left');
  const c9: IChildConfigMap = {
    0: {
      groupType: 'head',
      zIndex: 5,
    },
  };
  expect(getOffBorder(0, c9, true)).toBe('right');
  const c10: IChildConfigMap = {
    0: {
      groupType: 'group',
      zIndex: 5,
    },
  };
  expect(getOffBorder(0, c10, true)).toBe('right');
  const c11: IChildConfigMap = {
    0: {
      groupType: 'tail',
      zIndex: 5,
    },
  };
  expect(getOffBorder(0, c11, true)).toBe('no');
});
