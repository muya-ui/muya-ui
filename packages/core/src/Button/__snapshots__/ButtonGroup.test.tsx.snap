// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 ButtonGroup selected 快捷方式 1`] = `
.c2 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c7 + .c3 {
  margin-left: 12px;
}

.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 14px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  color: rgba(255,255,255,1);
  background-color: rgba(153,153,153,1);
  cursor: pointer;
  border-color: rgba(179,179,179,1);
  z-index: 2;
  border-width: 0;
  border-right-width: 1px;
  border-right-style: solid;
}

.c4:hover {
  color: rgba(255,255,255,1);
  background-color: rgba(179,179,179,1);
  border-color: rgba(179,179,179,1);
  z-index: 9;
}

.c4:active {
  color: rgba(255,255,255,1);
  background-color: rgba(140,140,140,1);
  border-color: rgba(179,179,179,1);
  z-index: 9;
}

.c5 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 0px;
  margin-left: -1px;
  color: rgba(255,255,255,1);
  background-color: rgba(26,122,248,1);
  cursor: pointer;
  border-color: rgba(51,139,255,1);
  z-index: 10;
  border-width: 0;
  border-right-width: 1px;
  border-right-style: solid;
  border-left-width: 1px;
  border-left-style: solid;
}

.c5:hover {
  color: rgba(255,255,255,1);
  background-color: rgba(51,139,255,1);
  border-color: rgba(51,139,255,1);
  z-index: 10;
}

.c5:active {
  color: rgba(255,255,255,1);
  background-color: rgba(12,108,235,1);
  border-color: rgba(51,139,255,1);
  z-index: 10;
}

.c6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 14px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  margin-left: -1px;
  color: rgba(255,255,255,1);
  background-color: rgba(153,153,153,1);
  cursor: pointer;
  border-color: rgba(179,179,179,1);
  z-index: 2;
  border-width: 0;
  border-left-width: 1px;
  border-left-style: solid;
}

.c6:hover {
  color: rgba(255,255,255,1);
  background-color: rgba(179,179,179,1);
  border-color: rgba(179,179,179,1);
  z-index: 9;
}

.c6:active {
  color: rgba(255,255,255,1);
  background-color: rgba(140,140,140,1);
  border-color: rgba(179,179,179,1);
  z-index: 9;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 + .c0 {
  margin-left: 8px;
}

.c8 + .c0 {
  margin-left: 12px;
}

<div
  className="c0 c1"
>
  <button
    className="c2 c3 c4"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
  <button
    className="c2 c3 c5"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
  <button
    className="c2 c3 c6"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
</div>
`;

exports[`测试 ButtonGroup 单节点 1`] = `
.c2 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 48px;
  padding: 0 23px 0 23px;
  font-weight: 600;
  font-size: 18px;
  line-height: 24px;
  width: 100%;
  border-radius: 2px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  border-style: solid;
  border-width: 1px;
}

.c4:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
}

.c4:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
}

.c4 + .c3 {
  margin-left: 12px;
}

.c5 + .c0 {
  margin-left: 8px;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 100%;
}

.c1 + .c0 {
  margin-left: 12px;
}

<div
  className="c0 c1"
>
  <button
    className="c2 c3 c4"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
</div>
`;

exports[`测试 ButtonGroup 子元素属性传递 1`] = `
.c2 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 48px;
  padding: 0 23px 0 23px;
  font-weight: 600;
  font-size: 18px;
  line-height: 24px;
  width: 100%;
  border-radius: 0px;
  margin-left: -1px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  z-index: 3;
  border-style: solid;
  border-width: 1px;
}

.c3:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c3:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 48px;
  padding: 0 23px 0 23px;
  font-weight: 600;
  font-size: 18px;
  line-height: 24px;
  width: 100%;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  margin-left: -1px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  z-index: 3;
  border-style: solid;
  border-width: 1px;
}

.c4:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c4:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c5 + .c0 {
  margin-left: 8px;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 100%;
}

.c1 + .c0 {
  margin-left: 12px;
}

<div
  className="c0 c1"
>
  <button
    className="c2 c3"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
  <button
    className="c2 c3"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
  <button
    className="c2 c4"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
</div>
`;

exports[`测试 ButtonGroup 正常情况 1`] = `
.c2 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  z-index: 3;
  border-style: solid;
  border-width: 1px;
}

.c3:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c3:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  margin-left: -1px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  z-index: 3;
  border-style: solid;
  border-width: 1px;
}

.c4:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c4:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 + .c0 {
  margin-left: 8px;
}

<div
  className="c0 c1"
>
  <button
    className="c2 c3"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
  <button
    className="c2 c4"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    AAA
  </button>
</div>
`;

exports[`测试 ButtonGroup 空情况 1`] = `
.c2 + .c0 {
  margin-left: 8px;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 100%;
}

.c1 + .c0 {
  margin-left: 12px;
}

<div
  className="c0 c1"
/>
`;
